File: bin/xurl.pl
#!/usr/bin/perl -w -s
# extract urls
# "cause qxurl from perlfaq9 just wasn't good enough"
# Mostly stolen from manpage for HTML::TokeParser
#
# William McVey
# July 27, 2001

require 5.002;
use HTML::TokeParser;

$p = HTML::TokeParser->new(shift||*STDIN);

$v=0;
while (my $token = $p->get_tag("a")) {
	my $url = $token->[1]{href} || "-";
	my $text = $p->get_trimmed_text("/a");
	if ($v) { 
		print "$url\t$text\n"; 
	} else {
		print "$url\n";
	}
} 

File: bin/xurl.py
#!/usr/bin/env python

"""A program to do something

Detailed documentation
"""

__author__ =	"William McVey"
__date__ = 	"24 May, 2004"
__revision__ =	"$Id:$"


import os
import sys
import lxml.etree
import urlparse
import re

class HTML:
	def __init__(self, text):
		self.tree = lxml.etree.HTML(text)

	def urls(self):
		urls = []
		try:
			base = tree.find("base").attrib["href"]
		except:
			base = ""
		for anchor in self.tree.xpath("//a[@href]"):
			url = urlparse.urljoin(base, anchor.attrib["href"])
			urls.append(url)
		return urls

if __name__ == '__main__':
	import sys
	import os
	from optparse import OptionParser       # aka Optik

	Progname=os.path.basename(sys.argv[0])
	Usage="""\
%prog usage: XXX:
%prog usage: -h
%prog usage: -V 
"""
	optparser = OptionParser(usage=Usage, version="%prog: $Id:$" )
	# optparser.add_option("-v", "--verbose", dest = "verbose",
	#   action="store_true", help="be verbose")
	#optparser.add_option("-N", "--name", dest="var_n", 
	#  action= "store" | "append" | "store_true" | "store_false" 
	#  type = "int"
	#  default="foo", metavar="SOME_STRING", help="store a string")
	(options, params) = optparser.parse_args()

	p = HTML(sys.stdin.read())
	for url in p.urls():
		print url
